Docker inspect "AppArmorProfile" field now shows "docker-default" when AppArmor is enabled and no other profile was defined

RobSkye opened the pull request:
This pull request fixes #25935 showing "AppArmorProfile" in docker inspect filled with "docker-default" when AppArmor is enabled and no other profile was defined using --security-opt.

thaJeztah commented:
@RobSkye did you close this on purpose, or by accident?

RobSkye commented:
By Accident...laptops and beds are bad combination....I'm editing the description too.

thaJeztah commented:
GitHub buttons are also a disaster at times, haha, no worries, thanks!

thaJeztah commented:
Guess this fixes https://github.com/docker/docker/issues/25935

RobSkye commented:
Yes.

tonistiigi commented on daemon/container_linux.go:
```diff
@@ -0,0 +1,20 @@
+		return err
+	}
+	if daemon.apparmorEnabled && container.AppArmorProfile == "" {
+		container.AppArmorProfile = "docker-default"
```
Also need to handle `--privileged`. See https://github.com/docker/docker/blob/master/daemon/oci_linux.go#L709-L714

Even with `privileged` fixed this is a slight behavioral change. If apparmor is enabled after container was created this var will not change. I think if this is only called on inspecting single container we can lazy load the value like we do in container startup.

tonistiigi reviewed (commented):
Please squash the commits as well.

thaJeztah commented:
@RobSkye can you address @tonistiigi's comments, and squash your commits?

RobSkye commented:
With the latest push, if you change apparmor configuration and apparmor completely disappear from the dockerhost (I have some comments about this below) the AppArmorProfile is updated with the expected configuration but still there is some issues we have to address:

1. If you do an inspect after an apparmor change and before the start of the container, still shows the old AppArmorProfile.
2. If you create the container with apparmor disabled but with --security-opt apparmor=someprofile, when you enable apparmor in dockerhost, AppArmorProfile will be "docker-default".
3. Apparmor status is checked ONLY checking the existence of /sys/kernel/security/apparmor and is not a good method, you can have apparmor completely shutdown and still that directory exists. see https://github.com/docker/docker/blob/master/pkg/sysinfo/sysinfo_linux.go#L57
4. You can run a container with --security-opt apparmor=someprofile and someprofile will be set in AppArmorProfile whether or not exists as an valid apparmor profile.

I don't know if addressing this issues are in the scope of the main issue. I' have the feeling that implementing a real-time check of the security configuration of containers will require more than a "slight behavioral change". 

I have some ideas of adding security information in the "docker ps" command and a more detailed "docker security" showing all the security settings of a container.

I look forward to your feedback.

tonistiigi commented on daemon/container_linux.go:
```diff
@@ -0,0 +1,34 @@
+func (daemon *Daemon) fixSecurityOptions(container *container.Container) {
+	if daemon.apparmorEnabled && !container.HostConfig.Privileged {
+		if container.AppArmorProfile == "unconfined" {
+			container.AppArmorProfile = defaultApparmorProfile
```
Don't understand this. Why is unconfined profile switched back to default.

RobSkye replied:
That is actually an error. I was trying to restore "docker-default" to the unconfined containers if the containers were created after apparmor was enabled but i didn't notice the effect of enabling apparmor to the containers with --security-opt app-armor:unconfined.

I'm gonna think on how manage this correctly.

RobSkye commented:
I did a little refactor of the patch in the last commit. Now inspect always shows an empty AppArmorProfile when the container is stopped ( maybe empty is not the best value but this can be changed easily to undefined, unconfined or something more clarifying).

The correct value of AppArmorProfile is set at container startup using the actual status of apparmor, the privileged flag and the configured apparmor settings using --security-opt.

LK4D4 commented:
@RobSkye I'm not sure that I like an idea of clearing profile on stopped containers - it's breaking change, and indeed it might be useful for debugging. Otherwise, the patch is ok for me.

mlaventure commented:
ping @RobSkye , needs a rebase.

I personally don't think showing the previous apparmor profile when inspecting old containers is an issue to be fixed otherwise. It properly reflect how this container was run.

RobSkye commented:
Ok, i'll delete that part and rebase/commit.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "25935-show-apparmor-default-profile-in-docker-inspect" git@github.com:RobSkye/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353964288
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

Commit:
Added an apparmorEnabled boolean in the Daemon struct to indicate if AppArmor is enabled or not. It is set in NewDaemon using sysInfo information.

Signed-off-by: Roberto Muñoz Fernández <robertomf@gmail.com>

Added an apparmorEnabled boolean in the Daemon struct to indicate if AppArmor is enabled or not. It is set in NewDaemon using sysInfo information.

Signed-off-by: Roberto Muñoz Fernández <robertomf@gmail.com>

gofmt'd

Signed-off-by: Roberto Muñoz Fernández <robertomf@gmail.com>

change the function name to something more adequate and changed the behaviour to show empty value on an apparmor disabled system.

Signed-off-by: Roberto Muñoz Fernández <robertomf@gmail.com>

go fmt

Signed-off-by: Roberto Muñoz Fernández <robertomf@gmail.com>

thaJeztah commented:
ping @tonistiigi @LK4D4 PTAL what's the status on this one?

tonistiigi commented on daemon/container_linux.go:
```diff
@@ -0,0 +1,22 @@
+	"github.com/docker/docker/container"
+)
+
+func (daemon *Daemon) fixSecurityOptions(container *container.Container) error {
```
I'd call this `saveApparmorConfig`

tonistiigi commented on daemon/container_linux.go:
```diff
@@ -0,0 +1,22 @@
+	if err := parseSecurityOpt(container, container.HostConfig); err != nil {
+		return err
+	}
+	if daemon.apparmorEnabled && !container.HostConfig.Privileged {
```
This shows `unconfined` when apparmor is disabled.

RobSkye replied:
Yes, do you want to show "Disabled" instead?

tonistiigi replied:
I think empty would do. `unconfined` implies that special privileged profile was applied(at least to me).

LK4D4 commented:
ping @RobSkye 
Would you mind to fix Tonis' comments? Then we can merge

LK4D4 commented:
LGTM
ping @tonistiigi

tonistiigi reviewed (approved):
LGTM
